### PR TITLE
Implemented RKC

### PR DESCRIPTION
**Pull request by deeepeshthakur**
@ChrisRackauckas @YingboMa I tried to implement RKC from #2 but I'm getting errors than I tried but couldn't resolve and also I have some doubts about changing some constants. It's basically the Julia translation of the fortran code from the paper [RKC: An explicit solver for parabolic PDEs](https://ir.cwi.nl/pub/4764).

### CODE REVIEW

File: src/alg_utils.jl [hunk 1]
```diff
@@ -321,6 +322,7 @@ beta1_default(alg::DP5,beta2) = typeof(beta2)(1//alg_order(alg)) - 3beta2/4
 beta1_default(alg::DP5Threaded,beta2) = typeof(beta2)(1//alg_order(alg)) - 3beta2/4
 
 gamma_default(alg::OrdinaryDiffEqAlgorithm) = 9//10
+gamma_default(alg::RKC) = 8//10
```
Comment (deeepeshthakur) [inline src/alg_utils.jl]: @ChrisRackauckas @YingboMa Changed gamma to match the constants of the paper

File: src/integrators/integrator_utils.jl [hunk 2]
```diff
@@ -237,40 +237,45 @@ function stepsize_controller!(integrator,alg::Union{StandardControllerAlgs,
...
 end
 function step_reject_controller!(integrator,alg::Union{StandardControllerAlgs,
-                                 OrdinaryDiffEqNewtonAdaptiveAlgorithm{CS,AD,:Standard}}) where {CS, AD}
+  OrdinaryDiffEqNewtonAdaptiveAlgorithm{CS,AD,:Standard}}) where {CS, AD}
   integrator.dt = integrator.qold
 end
 
-function stepsize_controller!(integrator,
-                              alg::OrdinaryDiffEqNewtonAdaptiveAlgorithm{CS,AD,:Predictive}) where {CS, AD}
+const PredictiveControllerAlgs = Union{RKC}
```
Comment (deeepeshthakur) [inline src/integrators/integrator_utils.jl]: @ChrisRackauckas @YingboMa I had to use the Gustafsson predictive stepsize controller.

File: src/integrators/integrator_utils.jl [hunk 3]
```diff
@@ -237,40 +237,45 @@ function stepsize_controller!(integrator,alg::Union{StandardControllerAlgs,
...
   # Gustafsson predictive stepsize controller
 
   if iszero(integrator.EEst)
     q = inv(integrator.opts.qmax)
   else
     gamma = integrator.opts.gamma
-    if alg isa RadauIIA5
-      @unpack nl_iters = integrator.cache
-      @unpack max_iter = alg
+    if alg isa RKC
```
Comment (deeepeshthakur) [inline src/integrators/integrator_utils.jl]: @ChrisRackauckas @YingboMa Added this condition because earlier the calls were designed for Implicit methods but except choosing fac, rest of the mechanics was exactly same as described in the paper

**Comment by deeepeshthakur** [comment]
@ChrisRackauckas @YingboMa If it helps, the convergence results for test problems are
**Evaluated: 0.0 ≈ 2 (atol=0.1)
Evaluated: 5.225007222098151e-7 ≈ 2 (atol=0.1)**

**Comment by ChrisRackauckas** [comment]
You should change that because it's based on the stability which is different for the two methods

**Comment by deeepeshthakur** [comment]
> You should change that because it's based on the stability which is different for the two methods

Okay, I'll do that. Any suggestion on other issues?
Specially why `integrator.opts.adaptive` is false

**Comment by ChrisRackauckas** [comment]
Adaptiveity is disabled during convergence tests

**Comment by deeepeshthakur** [comment]
> Adaptiveity is disabled during convergence tests

So how can I enable it?

**Comment by YingboMa** [comment]
Convergence tests are done in fixed timestep.

**Comment by deeepeshthakur** [comment]
> Convergence tests are done in fixed timestep.

So, the difference in the evaluated order is not due because adaptivity is off?

**Comment by YingboMa** [comment]
Yes. Check how you apply the coefficients.

**Comment by ChrisRackauckas** [comment]
>So, the difference in the evaluated order is not because adaptivity is off?

Yes, it has nothing to do with time stepping adaptivity. The order being essentially zero is a sign that something is really off. What does the convergence plot look like?

**Comment by deeepeshthakur** [comment]
I'll check the coefficients again and because the order was coming out to be zero, I assumed I did something wrong and never plotted the convergence plot.

**Comment by deeepeshthakur** [comment]
@ChrisRackauckas @YingboMa I think I figured out what was going haywire.
This snippet is from Wikipedia. It shows how the Chebyshev polynomials are selected over different domains of `x`
image

So in the fortran code they have selected the second formula assuming that `x >= 1` 

**Snippet from the paper**
image

**My Code**
```
  w0 = 1.0 + 2.0/(13.0*(mdeg^2.0))
  temp1 = w0^2.0 - 1.0
  temp2 = sqrt(temp1)
  arg   = mdeg*log(w0 + temp2)
```
As you can see `arg` is equivalent to `x` and `arg < 1` in our test cases. So we are using the wrong formula. I thinking of implementing both these formulas 
<a href="https://www.codecogs.com/eqnedit.php?latex=\large&space;\mathbf{T_{s}&space;=&space;\left\{\begin{matrix}&space;\cosh(s\cosh^{-1}(x))\hspace{1cm}&space;1&space;\leq&space;x\\\hspace{1cm}\cos(s\cos^{-1}(x))\hspace{1cm}&space;-1&space;\leq&space;x&space;\leq&space;1&space;\end{matrix}\right.}" target="_blank"></a>
But that would be a little different from the original fortran implementation. Should I?

**Comment by deeepeshthakur** [comment]
@ChrisRackauckas

File: src/caches/rkc_caches.jl [hunk 4]
```diff
@@ -88,3 +88,39 @@ end
 function alg_cache(alg::ROCK4,u,rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,tTypeNoUnits,uprev,uprev2,f,t,dt,reltol,p,calck,::Type{Val{false}})
   ROCK4ConstantCache(real(uEltypeNoUnits), real(uEltypeNoUnits),u)
 end
+
+mutable struct RKCConstantCache{zType} <: OrdinaryDiffEqConstantCache
+  #to match the types to call maxeig!
+  zprev::zType
+  function RKCConstantCache(zprev)
```
Comment (YingboMa) [inline src/caches/rkc_caches.jl]: You don't need to write a constructor explicitly.

File: src/alg_utils.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/alg_utils.jl]: yeah that's fine

File: src/integrators/integrator_utils.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply src/integrators/integrator_utils.jl]: add `OrdinaryDiffEqNewtonAdaptiveAlgorithm{CS,AD,:Predictive}` to the union

File: src/integrators/integrator_utils.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/integrators/integrator_utils.jl]: that works

File: src/integrators/integrator_utils.jl [hunk 5]
```diff
@@ -479,8 +483,8 @@ end
 function reset_fsal!(integrator)
   # Under these condtions, these algorithms are not FSAL anymore
   if typeof(integrator.cache) <: OrdinaryDiffEqMutableCache ||
-     (typeof(integrator.cache) <: CompositeCache &&
-      typeof(integrator.cache.caches[1]) <: OrdinaryDiffEqMutableCache)
```
Comment (ChrisRackauckas) [inline src/integrators/integrator_utils.jl]: keep these kinds of changes in a different PR.

File: src/rkc_utils.jl [hunk 6]
```diff
@@ -126,27 +153,26 @@ function maxeig!(integrator, cache::OrdinaryDiffEqMutableCache)
   end
   return false
 end
-
 """
-    choosedeg!(cache) -> nothing
+choosedeg!(cache) -> nothing
```
Comment (YingboMa) [inline src/rkc_utils.jl]: Why the indentation change?

File: src/perform_step/rkc_perform_step.jl [hunk 7]
```diff
@@ -296,3 +296,169 @@ end
...
+      z2   = z1
+      z1   = z
+      dz2  = dz1
+      dz1  = dz
+      d2z2 = d2z1
+      d2z1 = d2z
+    end
+  end
+  # error estimate
+  tmp = 0.8*(uprev - u) + 0.4*dt*(fsalfirst + gprev)
```
Comment (ChrisRackauckas) [inline src/perform_step/rkc_perform_step.jl]: this only needs to be calculated if adaptive?

File: src/perform_step/rkc_perform_step.jl [hunk 8]
```diff
@@ -296,3 +296,169 @@ end
...
+  arg   = mdeg*log(w0 + temp2)
+  w1    = (sinh(arg)*temp1) / (cosh(arg)*mdeg*temp2 - w0*sinh(arg))
+  b1    = 1.0/((2.0*w0)^2.0)
+  b2    = b1
+
+  # stage-1
+  gprev2 = copy(uprev)
+  μs     = w1*b1
+  gprev  = uprev + dt*μs*fsalfirst
+  th2  = 0.0
```
Comment (ChrisRackauckas) [inline src/perform_step/rkc_perform_step.jl]: this code is not type safe. These all assume Float64. Use `zero(eltype(u))` and `one(eltype(u))` instead

File: src/rkc_utils.jl (hunk 6 above)
Comment (deeepeshthakur) [reply src/rkc_utils.jl]: I auto-indented it. I'll change it back.

Comment (ChrisRackauckas) [reply src/rkc_utils.jl]: PRs for fixing indentation are good, but muddling the diff so it's hard to see what you did in a PR that's about adding/changing an algorithm is bad.

File: src/caches/rkc_caches.jl (hunk 4 above)
Comment (deeepeshthakur) [reply src/caches/rkc_caches.jl]: But every constant cache has one!

File: src/rkc_utils.jl (hunk 6 above)
Comment (deeepeshthakur) [reply src/rkc_utils.jl]: Yeah, I'll make sure it never happens again.

File: src/caches/rkc_caches.jl (hunk 4 above)
Comment (YingboMa) [reply src/caches/rkc_caches.jl]:
```julia
julia> struct A{T}
           a::T
       end

julia> A(1)
A{Int64}(1)

julia> struct B{T}
           b::T
           B(x) = new{typeof(x)}(x)
       end

julia> B(1)
B{Int64}(1)
```
You don't need to add a trivial constructor. Julia adds it automatically.

**Comment by ChrisRackauckas** [comment]
The constructor issue is the last thing?

**Comment by deeepeshthakur** [comment]
Yes. I made all the other corrections.
